device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttIotHubConnection.java

```diff
@@ -273,6 +251,27 @@ public void open() throws TransportException
 
             log.debug("Opening MQTT connection...");
 
+            if (this.config.getSasTokenAuthentication() != null)
+            {
+                try
+                {
+                    log.trace("Setting password for MQTT connection since it is a SAS token authenticated connection");
+                    this.deviceMessaging.updatePassword(this.config.getSasTokenAuthentication().getSasToken());
+                }
+                catch (IOException e)
+                {
+                    throw new TransportException("Failed to open the MQTT connection because a SAS token could not be retrieved", e);
+                }
+            }
+
+            // MqttAsyncClient's are unusable after they have been closed. This logic creates a new client
+            // each time an open is called
+            MqttAsyncClient mqttAsyncClient = buildMqttAsyncClient(this.serverUri, this.clientId);
+            mqttAsyncClient.setCallback(this.deviceMessaging);
+            this.deviceMessaging.setMqttAsyncClient(mqttAsyncClient);
+            this.deviceTwin.setMqttAsyncClient(mqttAsyncClient);
+            this.deviceMethod.setMqttAsyncClient(mqttAsyncClient);
```

Comment:
As noted above, we only assign the (single use) mqtt async client at open time, so there is no risk of an mqtt async client calling open again after it was already closed